### ISSUE TITLE
row_cache_test.cc:test_exception_safety_of_reads is flaky

### ISSUE THREAD

**Issue by haaawk** [scylladb/scylla#5642]
The test fails randomly from time to time:

Here's the summary of the log from one of the failures (full log in the attachment -> [test.txt](https://github.com/scylladb/scylla/files/4107334/test.txt)):

```
test/boost/row_cache_test.cc(0): Entering test case "test_exception_safety_of_reads"
random_mutation_generator seed: 4120932408
./test/lib/flat_mutation_reader_assertions.hh(345): fatal error: in "test_exception_safety_of_reads": Expected { <lots of data goes here> }, but got end of stream, at:   0x1cc0b99
  0x932f65
  0x8c3882
  0x8c4209
  0x1cc0845
  0x8c3f15
  0x8c4124
  0x1c029e2

test/boost/row_cache_test.cc(0): Leaving test case "test_exception_safety_of_reads"; testing time: 2137836us
```

It might be related to #3095

**Comment by haaawk** [comment]
FYI @tgrabiec @penberg @hagitsegev

**Comment by bhalevy** [comment]
@tgrabiec I see that you wrote this test case.
can you please look into this issue?

**Comment by bhalevy** [comment]
Seen in https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/1546/testReport/junit/test_boost_memtable_test/cc/test_exception_safety_of_partition_range_reads/

**Comment by denesb** [comment]
Most likely uncovered by https://github.com/scylladb/scylla/commit/e8a948ece62c0cc9767ead4a2b7e9bb71903b4c3.

**Comment by avikivity** [comment]
Perhaps we should temporarily disable it in row_cache_test (keeping the bug open), since it's not a regression. Provided we can prove this is the immediate cause.

**Comment by denesb** [comment]
There is no nice way of disabling this for one test, other than commenting the lines [which triggers the error injection](https://github.com/scylladb/scylla/blob/e1b22b6a4c56b4f1d0adf65d1a11db4bcb51fe7d/test/boost/row_cache_test.cc#L2347).

To turn the infrastructure on/off the entire codebase has to be recompiled.

**Comment by avikivity** [comment]
Well, that was the whole point of the test. So it ran without testing anything?

**Comment by denesb** [comment]
Before https://github.com/scylladb/scylla/commit/e8a948ece62c0cc9767ead4a2b7e9bb71903b4c3 yes, and all the other similar tests too (we actually have quite a few).

**Comment by slivne** [comment]
@denesb is the test running now ?

**Comment by denesb** [comment]
@slivne yes, failure injection is enabled in debug and dev modes.

**Comment by bhalevy** [comment]
Still happening...
https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/2074/testReport/junit/test_boost_row_cache_test/cc/test_exception_safety_of_reads/

**Comment by espindola** [comment]
With a41258b1160d3cef24935425a89ea1479f974b61 and the attached change I can reproduce it every time with

./build/dev/test/boost/row_cache_test_g -t test_exception_safety_of_reads --report_level=no  --catch_system_errors=no  -- -c2 -m2G --overprovisioned --unsafe-bypass-fsync 1 --blocked-reactor-notify-ms 2000000 --collectd 0 --random-seed 404629206

[t.patch.txt](https://github.com/scylladb/scylla/files/4754192/t.patch.txt)

**Comment by espindola** [comment]
I was able to reduce the testcase a bit and pushed it to

https://github.com/espindola/scylla/tree/debug-next-failure

I still don't understand why, but the failure happens if the cache is evicted during two very specific allocations. I have attached the backtrace of both.

To reproduce, run
 ./build/dev/test/boost/row_cache_test_g -t test_exception_safety_of_reads --report_level=no --catch_system_errors=no -- -c2 -m2G --overprovisioned --unsafe-bypass-fsync 1 --blocked-reactor-notify-ms 2000000 --collectd 0 --random-seed 404629206

[bt1.txt](https://github.com/scylladb/scylla/files/4755582/bt1.txt)
[bt2.txt](https://github.com/scylladb/scylla/files/4755583/bt2.txt)

I would appreciate If someone more familiar with the cache could take it from here.

**Comment by avikivity** [comment]
@tgrabiec that's you

**Comment by bhalevy** [comment]
Cc @xemul

**Comment by tgrabiec** [comment]
This is caused by the fact that (1) key comparison may involve linearization if the key is large, which may need to reclaim, which invalidates iterators (and linearization_context), and (2) the code is unprepared for this (in many places) and we hit undefined behavior.

Refs https://github.com/scylladb/scylla/issues/6108

**Comment by tgrabiec** [comment]
Fixed by e81fc1f
